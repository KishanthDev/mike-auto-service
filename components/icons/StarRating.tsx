'use client';

import { FC } from "react";
import { Star, StarHalf } from "lucide-react";

interface StarRatingProps {
  rating: number; // e.g. 3.5
}

const StarRating: FC<StarRatingProps> = ({ rating }) => {
  const fullStars = Math.floor(rating);
  const hasHalfStar = rating % 1 >= 0.5;
  const emptyStars = 5 - fullStars - (hasHalfStar ? 1 : 0);

  return (
    <div className="flex items-center gap-1 mt-1">
      {/* Full Stars */}
      {Array(fullStars)
        .fill(0)
        .map((_, i) => (
          <Star key={`full-${i}`} className="w-5 h-5 fill-yellow-500 text-yellow-500" />
        ))}

      {/* Half Star */}
      {hasHalfStar && (
        <StarHalf className="w-5 h-5 fill-yellow-500 text-yellow-500" />
      )}

      {/* Empty Stars */}
      {Array(emptyStars)
        .fill(0)
        .map((_, i) => (
          <Star key={`empty-${i}`} className="w-5 h-5 text-yellow-500" />
        ))}

      {/* Rating Number */}
      <span className="ml-2 text-sm font-medium text-gray-800 dark:text-gray-300">
        {rating.toFixed(1)}
      </span>
    </div>
  );
};

export default StarRating;
