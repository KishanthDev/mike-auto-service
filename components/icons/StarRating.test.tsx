import React from 'react';
import { render, screen } from '@testing-library/react';
import StarRating from './StarRating';

jest.mock('lucide-react', () => ({
  Star: jest.fn((props) => <svg data-testid="star" {...props} />),
  StarHalf: jest.fn((props) => <svg data-testid="star-half" {...props} />),
}));

describe('StarRating Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correct number of stars for integer ratings', () => {
    render(<StarRating rating={3} />);
    
    const stars = screen.getAllByTestId('star');
    expect(stars).toHaveLength(5); // 5 total stars (filled + empty)
    
    stars.slice(0, 3).forEach(star => {
      expect(star).toHaveClass('fill-yellow-500');
    });
    
    stars.slice(3).forEach(star => {
      expect(star).not.toHaveClass('fill-yellow-500');
    });
    
    expect(screen.queryByTestId('star-half')).not.toBeInTheDocument();
    
    expect(screen.getByText('3.0')).toBeInTheDocument();
  });

  it('renders half star for .5 ratings', () => {
    render(<StarRating rating={3.5} />);
    
    const fullStars = screen.getAllByTestId('star')
      .filter(star => star.classList.contains('fill-yellow-500'));
    expect(fullStars).toHaveLength(3);
    
    const halfStar = screen.getByTestId('star-half');
    expect(halfStar).toBeInTheDocument();
    expect(halfStar).toHaveClass('fill-yellow-500');
    
    const emptyStars = screen.getAllByTestId('star')
      .filter(star => !star.classList.contains('fill-yellow-500'));
    expect(emptyStars).toHaveLength(1);
    
    // Check rating number
    expect(screen.getByText('3.5')).toBeInTheDocument();
  });

  it('applies correct styling classes', () => {
    render(<StarRating rating={2.5} />);
    
    const stars = screen.getAllByTestId('star');
    stars.slice(0, 2).forEach(star => {
      expect(star).toHaveClass('fill-yellow-500');
      expect(star).toHaveClass('text-yellow-500');
    });
    
    stars.slice(3).forEach(star => {
      expect(star).toHaveClass('text-yellow-500');
      expect(star).not.toHaveClass('fill-yellow-500');
    });
    
    const halfStar = screen.getByTestId('star-half');
    expect(halfStar).toHaveClass('fill-yellow-500');
    expect(halfStar).toHaveClass('text-yellow-500');
    
    const ratingText = screen.getByText('2.5');
    expect(ratingText).toHaveClass('text-gray-800');
    expect(ratingText).toHaveClass('dark:text-gray-300');
  });

  it('matches snapshot for various ratings', () => {
    const { asFragment: fragment1 } = render(<StarRating rating={0} />);
    expect(fragment1()).toMatchSnapshot();
    
    const { asFragment: fragment2 } = render(<StarRating rating={3.5} />);
    expect(fragment2()).toMatchSnapshot();
    
    const { asFragment: fragment3 } = render(<StarRating rating={5} />);
    expect(fragment3()).toMatchSnapshot();
  });
});