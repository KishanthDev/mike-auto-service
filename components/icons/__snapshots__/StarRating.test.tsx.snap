// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRating Component matches snapshot for various ratings 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-1 mt-1"
  >
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <span
      class="ml-2 text-sm font-medium text-gray-800 dark:text-gray-300"
    >
      0.0
    </span>
  </div>
</DocumentFragment>
`;

exports[`StarRating Component matches snapshot for various ratings 2`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-1 mt-1"
  >
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star-half"
    />
    <svg
      class="w-5 h-5 text-yellow-500"
      data-testid="star"
    />
    <span
      class="ml-2 text-sm font-medium text-gray-800 dark:text-gray-300"
    >
      3.5
    </span>
  </div>
</DocumentFragment>
`;

exports[`StarRating Component matches snapshot for various ratings 3`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-1 mt-1"
  >
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <svg
      class="w-5 h-5 fill-yellow-500 text-yellow-500"
      data-testid="star"
    />
    <span
      class="ml-2 text-sm font-medium text-gray-800 dark:text-gray-300"
    >
      5.0
    </span>
  </div>
</DocumentFragment>
`;
